### PR TITLE
Mepts 290 common queries patients on tb

### PR DESCRIPTION
**Pull request by ItsMurumba**
**Added query to get all Patients on TB Treatment:**

1. All patients who in “Patient Clinical Record of ART - Ficha de Seguimento (adults and children)” have at least TB Treatment (Tratamento de TB) Start Date (Data de Início) and the TB end date (Data do Fim) is not registered or its after the reporting end date OR

2. Include all patients who are enrolled in TB Program with enrollment Date (Data de Admissão) and the TB end date (Data do Fim) is not registered or its after the reporting end date OR

3. Patients marked last “Diagnóstico TB activo – SIM” – Ficha Clinica Master Card in the last 6 months OR

4. Patients marked last “Tratamento TB– Inicio (I) or Continua (c) ” – Ficha Clinica MasterCard Patients on treatment for TB.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java [hunk 1]
```diff
@@ -336,4 +337,32 @@ private CohortDefinition getPatientsTransferredOutOrSuspended(
...
+
+  /**
+   * Get All Patients On TB Treatment
+   *
+   * @return
+   */
+  public CohortDefinition getPatientsOnTbTreatment() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("patientsOnTbTreatment");
+    cd.addParameter(new Parameter("onOrBefore", "onOrBefore", Date.class));
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: The parameters don't match. The SQL query expects `startDate`, `endDate` and `location`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            + "         ON p.patient_id = e.patient_id "
+            + "       JOIN obs o "
+            + "         ON p.patient_id = o.person_id "
+            + "WHERE  e.encounter_type IN ( %d, %d ) "
+            + "       AND p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.concept_id = %d "
+            + "       AND o.value_datetime > :endDate "
+            + "       AND e.location_id = :location "
+            + "UNION "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: Looking at the [ADULTO: SEGUIMENTO form](http://thebugatti.e-saude.net:8080/fghmaster/module/htmlformentry/htmlForm.form?id=280), I think `TB Start Date` and `TB End Date` should belong to the same encounter to be valid (please confirm with Pinki).
Having 3 unions don't guarantee that. It should look something like this:
```
SELECT p.patient_id
FROM   patient p
           JOIN encounter e
                ON p.patient_id = e.patient_id
           JOIN obs start
                ON e.encounter_id = start.encounter_id
           JOIN obs end
                ON e.encounter_id = end.encounter_id
WHERE  start.concept_id = 1113
  AND start.value_datetime BETWEEN :startDate AND :endDate
  AND end.concept_id = 6120
  AND ( end.value_datetime IS NULL
    OR end.value_datetime > :endDate )
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java [hunk 3]
```diff
@@ -336,4 +337,32 @@ private CohortDefinition getPatientsTransferredOutOrSuspended(
...
+    cd.addParameter(new Parameter("location", "location", Location.class));
+
+    cd.setQuery(
+        HivQueries.getPatientsOnTbTreatmentQuery(
+            hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId(),
+            hivMetadata.getARVPediatriaSeguimentoEncounterType().getEncounterTypeId(),
+            hivMetadata.getTBDrugStartDateConcept().getConceptId(),
+            hivMetadata.getTBDrugEndDateConcept().getConceptId(),
+            hivMetadata.getTBProgram().getProgramId(),
+            hivMetadata.getActiveOnProgramConcept().getConceptId(),
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: I think you forgot change to the `ACTIVE ON PROGRAM` state from `STATE TB` workflow.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            + "       JOIN patient_state ps "
+            + "         ON pp.patient_program_id = ps.patient_program_id "
+            + "WHERE  pp.program_id = %d "
+            + "       AND ps.state = %d "
+            + "       AND e.location_id = :location "
+            + "       AND o.concept_id = %d "
+            + "       AND o.value_datetime > :endDate "
+            + "       AND p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "UNION "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: `Start date` and `End date` refer to the patient state. It is not necessary to join with `encounter` and `obs`.
```
WHERE ps.start_date BETWEEN :startDate AND :endDate
  AND (ps.end_date IS NULL OR ps.end_date > :endDate)
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            + "         ON p.patient_id = o.person_id "
+            + "WHERE  e.encounter_type IN ( %d, %d ) "
+            + "       AND p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.concept_id = %d "
+            + "       AND o.value_coded = %d "
+            + "       AND e.encounter_datetime BETWEEN :endDate AND Date_sub(:endDate, "
+            + "                                                     INTERVAL 6 month) "
+            + "       AND e.location_id = :location "
+            + "UNION "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: The query should be looking for the **Last** `Diagnóstico TB activo – SIM`. 

You would need an inner query to fetch the last encounter with that obs, then check if the `encounter_datetime` falls `BETWEEN Date_sub(:endDate, INTERVAL 6 MONTH) AND :endDate`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            + "         ON p.patient_id = e.patient_id "
+            + "       JOIN obs o "
+            + "         ON p.patient_id = o.person_id "
+            + "WHERE  e.encounter_type IN ( %d, %d ) "
+            + "       AND p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.concept_id = %d "
+            + "       AND o.value_coded IN ( %d, %d ) "
+            + "       AND o.value_datetime BETWEEN :startDate AND :endDate "
+            + "       AND e.location_id = :location ";
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: Same as previous comment, we need the **Last** `Tratamento TB– Inicio (I) or Continua (c)`.

**Review by ynurmahomed — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 1 above)
Comment (ItsMurumba) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: Done.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 3 above)
Comment (ItsMurumba) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: Done.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java (hunk 2 above)
Comment (ItsMurumba) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: Made changes after consultation with **Pinki**.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java (hunk 5 above)
Comment (ItsMurumba) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/HivQueries.java]: Done.

**Pull request merged**